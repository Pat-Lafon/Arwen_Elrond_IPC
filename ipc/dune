(library
 (name ipc)
 (package my_project)
 (preprocess
  (pps ppx_deriving_yojson))
 (libraries inference language yojson))
