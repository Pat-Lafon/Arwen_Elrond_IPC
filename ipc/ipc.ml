open Yojson

(* https://stackoverflow.com/questions/70917952/how-can-i-change-how-ppx-yojson-conv-represents-variants *)
let rec to_assoc : [> Safe.t ] -> Safe.t as 'a = function
  (* todo is this true only for strings or also other things? *)
  | `List [ `String "Pair"; `List [ key; value ] ] ->
      `Assoc [ ("Pair", `List [ key; to_assoc value ]) ]
  | `List (`String "Pair" :: _) -> failwith "Not possible"
  | `List (`String key :: []) -> `String key
  | `List [ `String key; elem ] -> `Assoc [ (key, to_assoc elem) ]
  | `List (`String key :: l) -> `Assoc [ (key, `List (List.map to_assoc l)) ]
  | `List l -> `List (List.map to_assoc l)
  | `Assoc [ (key, l) ] -> `Assoc [ (key, to_assoc l) ]
  | x -> x

(* todo does this need to be rec? *)
let rec to_list = function
  | `Assoc [ (key, `List l) ] -> `List (`String key :: l)
  | `Assoc [ (key, `String l) ] -> `List [ `String key; `String l ]
  | `Assoc [ (key, l) ] -> `List [ `String key; l ]
  | `String key -> `List (`String key :: [])
  | x -> x

type elrondValue =
  | L of int list
  (* | T of int Utils.Tree.t *)
  | I of int
  | B of bool
  (* | TI of (int, int) Utils.LabeledTree.t
     | TB of (int, bool) Utils.LabeledTree.t *)
  | NotADt
[@@deriving yojson]

type elrondLiteral = Int of int | Bool of bool | IntList of int list
(*   | IntTree of int Utils.Tree.t *)
[@@deriving yojson]

type elrondType = Bool | Int | IntList | IntTree | IntTreeI | IntTreeB
[@@deriving yojson]

type elrondSimpleExpr =
  | Literal of elrondType * elrondLiteral
  | Var of elrondType * string
  | Op of elrondType * string * elrondSimpleExpr list
[@@deriving yojson]

type elrondPred =
  | True
  | Atom of elrondSimpleExpr
  | Implies of elrondPred * elrondPred
  | Ite of elrondPred * elrondPred * elrondPred
  | Not of elrondPred
  | And of elrondPred list
  | Or of elrondPred list
  | Iff of elrondPred * elrondPred
[@@deriving yojson]

type free_var = elrondType * string [@@deriving yojson]
type elrondTpedvar = elrondType * string [@@deriving yojson]
type elrondForallFormula = free_var list * elrondPred [@@deriving yojson]
type elrondSpec = elrondTpedvar list * elrondForallFormula [@@deriving yojson]

type elrondResult =
  | Cex of (elrondValue * string) list list
  | Result of (elrondSpec * string) list
[@@deriving yojson]

type elrondMessage = Message of string | Result of elrondResult
[@@deriving yojson]

type elrondPredicates = string list [@@deriving yojson]

type arwenSetup = {
  sourcefile : string;
  assertionfile : string;
  outputdir : string;
  client_name : string;
  predicates : elrondPredicates;
}
[@@deriving yojson]

type arwenMessage = Start | Test | Message of string | Setup of arwenSetup
[@@deriving yojson]

let dbg yojson =
  Printf.eprintf "Debuged yojson: %s\n" (Yojson.Safe.show yojson);
  yojson

let arwenMessage_to_yojson x = to_assoc (arwenMessage_to_yojson x)
let arwenMessage_of_yojson x = arwenMessage_of_yojson (to_list x)
let elrondMessage_to_yojson x = elrondMessage_to_yojson x |> to_assoc
let elrondMessage_of_yojson x = elrondMessage_of_yojson (to_list x)
let elrondValue_to_yojson x = elrondValue_to_yojson x |> to_assoc
let elrondValue_of_yojson x = elrondValue_of_yojson (to_list x)
let elrondLiteral_to_yojson x = elrondLiteral_to_yojson x |> to_assoc
let elrondLiteral_of_yojson x = elrondLiteral_of_yojson (to_list x)
let elrondType_to_yojson x = elrondType_to_yojson x |> to_assoc
let elrondType_of_yojson x = elrondType_of_yojson (to_list x)
let elrondSimpleExpr_to_yojson x = elrondSimpleExpr_to_yojson x |> to_assoc
let elrondSimpleExpr_of_yojson x = elrondSimpleExpr_of_yojson (to_list x)
let elrondPred_to_yojson x = elrondPred_to_yojson x |> to_assoc
let elrondPred_of_yojson x = elrondPred_of_yojson (to_list x)
let free_var_to_yojson x = free_var_to_yojson x |> to_assoc
let free_var_of_yojson x = free_var_of_yojson (to_list x)
let elrondTpedvar_to_yojson x = elrondTpedvar_to_yojson x |> to_assoc
let elrondTpedvar_of_yojson x = elrondTpedvar_of_yojson (to_list x)

let elrondForallFormula_to_yojson x =
  elrondForallFormula_to_yojson x |> to_assoc

let elrondForallFormula_of_yojson x = elrondForallFormula_of_yojson (to_list x)
let elrondSpec_to_yojson x = elrondSpec_to_yojson x |> to_assoc
let elrondSpec_of_yojson x = elrondSpec_of_yojson (to_list x)

(* todo maybe no immediate conversions here because of list of string pairs? *)
(* Or we could explicitly tag these pairs as enum pairs *)
let elrondResult_to_yojson x = elrondResult_to_yojson x |> to_assoc
let elrondResult_of_yojson x = elrondResult_of_yojson (to_list x)

let into_elrondValue (v : Pred.Value.t) : elrondValue =
  match v with
  | L l -> L l
  | I i -> I i
  | B b -> B b
  | NotADt -> NotADt
  | _ -> failwith "into_elrondValue: not implemented"

let into_str_val_list (m : Pred.Value.t Utils.StrMap.t) :
    (elrondValue * string) list =
  Utils.StrMap.fold (fun k v acc -> (into_elrondValue v, k) :: acc) m []

let into_elrondType (t : Tp.Tp.t) : elrondType =
  match t with
  | Tp.Tp.Bool -> Bool
  | Tp.Tp.Int -> Int
  | Tp.Tp.IntList -> IntList
  | Tp.Tp.IntTree -> IntTree
  | Tp.Tp.IntTreeI -> IntTreeI
  | Tp.Tp.IntTreeB -> IntTreeB

let into_elrondTpedvar (t : Tp.Tp.tpedvar) : elrondTpedvar =
  let tp, v = t in
  (into_elrondType tp, v)

let into_elrondLit (l : Inference.SingleAbd.SE.L.t) : elrondLiteral =
  match l with
  | Int i -> Int i
  | Bool b -> Bool b
  | IntList l -> IntList l
  | IntTree _t -> (* IntTree t *) failwith "into_elrondLit: not implemented"

let rec into_elrondSimpleExpr (e : Inference.SingleAbd.SE.t) : elrondSimpleExpr
    =
  match e with
  | Literal (t, l) -> Literal (into_elrondType t, into_elrondLit l)
  | Var (t, v) -> Var (into_elrondType t, v)
  | Op (t, op, es) ->
      Op (into_elrondType t, op, List.map into_elrondSimpleExpr es)

let rec into_elrondPred (p : Language.SpecAst.E.t) : elrondPred =
  match p with
  | True -> True
  | Atom e -> Atom (into_elrondSimpleExpr e)
  | Implies (p1, p2) -> Implies (into_elrondPred p1, into_elrondPred p2)
  | Ite (p1, p2, p3) ->
      Ite (into_elrondPred p1, into_elrondPred p2, into_elrondPred p3)
  | Not p -> Not (into_elrondPred p)
  | And ps -> And (List.map into_elrondPred ps)
  | Or ps -> Or (List.map into_elrondPred ps)
  | Iff (p1, p2) -> Iff (into_elrondPred p1, into_elrondPred p2)

let into_elrondFreeVar (fv : Language.Epr.free_variable) : free_var =
  let t, v = fv in
  (into_elrondType t, v)

let into_elrondForallFormula (f : Language.SpecAst.E.forallformula) :
    elrondForallFormula =
  let fv, p = f in
  (List.map into_elrondFreeVar fv, into_elrondPred p)

let into_elrondSpec (s : Language.SpecAst.spec) : elrondSpec =
  let fv, f = s in
  (List.map into_elrondTpedvar fv, into_elrondForallFormula f)

let into_elrondResult (mii : Inference.SpecAbduction.multi_infer_result) :
    elrondResult =
  match mii with
  | Cex cex -> Cex (List.map into_str_val_list cex)
  | Result res ->
      Result
        (Utils.StrMap.fold
           (fun k v acc -> (into_elrondSpec v, k) :: acc)
           res [])

let into_elrondMessage (elrondResult : elrondResult) : elrondMessage =
  Result elrondResult
