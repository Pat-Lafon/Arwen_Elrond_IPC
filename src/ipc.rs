use std::fmt::Display;

use itertools::Itertools;
use serde::{Deserialize, Serialize};

use crate::ipc_assertion::AssertionType;

#[derive(Serialize, Deserialize, Debug)]
pub struct ArwenSetup {
    pub sourcefile: String,
    pub assertionfile: String,
    pub outputdir: String,
    pub client_name: String,
    pub predicates: ElrondPredicates,
}

#[derive(Serialize, Deserialize, Debug)]
pub enum ArwenMessage {
    Start,
    Test,
    Setup(ArwenSetup),
    Message(String),
}

#[derive(Serialize, Deserialize, PartialEq, Debug)]
pub enum ElrondMessage {
    Message(String),
    Result(ElrondResult),
}

impl Display for ElrondMessage {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            ElrondMessage::Message(msg) => write!(f, "Message: {}", msg),
            ElrondMessage::Result(result) => write!(f, "Result:\n{}", result),
        }
    }
}

#[derive(Serialize, Deserialize, PartialEq, Debug)]
pub enum ElrondValue {
    L(Vec<i64>),
    I(i64),
    B(bool),
    NotADt,
}

impl Display for ElrondValue {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            ElrondValue::L(l) => write!(f, "[{}]", l.iter().join(", ")),
            ElrondValue::I(i) => write!(f, "{}", i),
            ElrondValue::B(b) => write!(f, "{}", b),
            ElrondValue::NotADt => write!(f, "NotADt"),
        }
    }
}

#[derive(Serialize, Deserialize, PartialEq, Debug)]
pub enum ElrondLiteral {
    Int(i64),
    Bool(bool),
    IntList(Vec<i64>),
}

impl Display for ElrondLiteral {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            ElrondLiteral::Int(i) => write!(f, "{i}"),
            ElrondLiteral::Bool(b) => write!(f, "{b}"),
            ElrondLiteral::IntList(l) => write!(f, "[{}]", l.iter().join(", ")),
        }
    }
}

#[derive(Serialize, Deserialize, PartialEq, Debug, Clone)]
pub enum ElrondType {
    Bool,
    Int,
    IntList,
    IntTree,
    IntTreeI,
    IntTreeB,
    Tuple(Vec<ElrondType>),
    Generic(String),
    Arrow(Box<ElrondType>, Box<ElrondType>),
}

impl Display for ElrondType {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            ElrondType::Bool => write!(f, "bool"),
            ElrondType::Int => write!(f, "int"),
            ElrondType::IntList => write!(f, "int list"),
            ElrondType::IntTree => write!(f, "int tree"),
            ElrondType::IntTreeI => write!(f, "int treei"),
            ElrondType::IntTreeB => write!(f, "int treeb"),
            ElrondType::Generic(s) => write!(f, "{s}"),
            ElrondType::Tuple(v) => {
                write!(f, "({})", v.iter().map(ToString::to_string).join(", "))
            }
            ElrondType::Arrow(t1, t2) => write!(f, "({t1} -> {t2})"),
        }
    }
}

#[derive(Serialize, Deserialize, PartialEq, Debug)]
pub enum ElrondSimpleExpr {
    Literal(ElrondType, ElrondLiteral),
    Var(ElrondType, String),
    Op(ElrondType, String, Vec<ElrondSimpleExpr>),
    Tuple(Vec<ElrondSimpleExpr>),
}

impl Display for ElrondSimpleExpr {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            ElrondSimpleExpr::Literal(_ty, lit) => write!(f, "{lit}"),
            ElrondSimpleExpr::Var(_ty, name) => write!(f, "{name}"),
            ElrondSimpleExpr::Op(_ty, op, args) if op == "<=" || op == ">=" => {
                write!(f, "({} {} {})", args[0], op, args[1])
            }
            ElrondSimpleExpr::Op(_ty, op, args) => {
                write!(f, "{op} {}", args.iter().map(ToString::to_string).join(" "))
            }
            ElrondSimpleExpr::Tuple(t) => {
                write!(f, "({})", t.iter().map(ToString::to_string).join(", "))
            }
        }
    }
}

#[derive(Serialize, Deserialize, PartialEq, Debug)]
pub enum ElrondPred {
    True,
    Atom(ElrondSimpleExpr),
    Implies(Box<ElrondPred>, Box<ElrondPred>),
    Ite(Box<ElrondPred>, Box<ElrondPred>, Box<ElrondPred>),
    Not(Box<ElrondPred>),
    And(Vec<ElrondPred>),
    Or(Vec<ElrondPred>),
    Iff(Box<ElrondPred>, Box<ElrondPred>),
}

impl ElrondPred {
    pub fn size(&self) -> u32 {
        match self {
            ElrondPred::True => 1,
            ElrondPred::Atom(_) => 1,
            ElrondPred::Iff(p1, p2) | ElrondPred::Implies(p1, p2) => 1 + p1.size() + p2.size(),
            ElrondPred::Ite(p1, p2, p3) => 1 + p1.size() + p2.size() + p3.size(),
            ElrondPred::Not(p) => 1 + p.size(),
            ElrondPred::And(p_vec) | ElrondPred::Or(p_vec) => {
                1 + p_vec.iter().map(ElrondPred::size).sum::<u32>()
            }
        }
    }
}

impl Display for ElrondPred {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            ElrondPred::True => write!(f, "true"),
            ElrondPred::Atom(expr) => write!(f, "{expr}"),
            ElrondPred::Implies(lhs, rhs) => write!(f, "implies ({lhs}) ({rhs})"),
            ElrondPred::Ite(cond, lhs, rhs) => write!(f, "(if {cond} then {lhs} else {rhs})"),
            ElrondPred::Not(pred) => write!(f, "(!{pred})"),
            ElrondPred::And(preds) => write!(
                f,
                "({})",
                preds.iter().map(|p| format!("{}", p)).join(" && ")
            ),
            ElrondPred::Or(preds) => write!(
                f,
                "({})",
                preds.iter().map(|p| format!("{}", p)).join(" || ")
            ),
            ElrondPred::Iff(lhs, rhs) => write!(f, "iff ({lhs}) ({rhs})"),
        }
    }
}

pub enum ElrondOp {
    Eq,
    Le,
    Ge,
    Ne,
    Lt,
    Gt,
}

#[derive(Serialize, Deserialize, PartialEq, Debug)]
pub struct FreeVar(pub ElrondType, pub String);

impl Display for FreeVar {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let FreeVar(ty, name) = self;
        write!(f, "{name} : {ty}")
    }
}

#[derive(Serialize, Deserialize, PartialEq, Debug)]
pub struct ElrondTpedvar(pub AssertionType, pub String);

impl Display for ElrondTpedvar {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let ElrondTpedvar(ty, name) = self;
        write!(f, "{name} : {ty}")
    }
}

#[derive(Serialize, Deserialize, PartialEq, Debug)]
pub struct ElrondForallFormula(pub Vec<FreeVar>, pub ElrondPred);

impl Display for ElrondForallFormula {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let ElrondForallFormula(vars, formula) = self;
        if vars.is_empty() {
            write!(f, "{formula}")
        } else {
            write!(
                f,
                "forall {} . {formula}",
                (vars.iter().map(ToString::to_string).join(","))
            )
        }
    }
}

#[derive(Serialize, Deserialize, PartialEq, Debug)]
pub struct ElrondSpec(pub Vec<ElrondTpedvar>, pub ElrondForallFormula);

impl Display for ElrondSpec {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let ElrondSpec(vars, formula) = self;
        if vars.is_empty() {
            write!(f, "{formula}")
        } else {
            write!(
                f,
                "{} ‚ä¢ {formula}",
                (vars.iter().map(ToString::to_string).join(","))
            )
        }
    }
}

#[derive(Serialize, Deserialize, PartialEq, Debug)]
pub enum ElrondResult {
    Cex(Vec<Vec<(ElrondValue, String)>>),
    Result(Vec<(ElrondSpec, String)>),
}

impl Display for ElrondResult {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            ElrondResult::Cex(c) => {
                for (i, cex) in c.iter().enumerate() {
                    writeln!(f, "Cex {i}:")?;
                    for (val, name) in cex {
                        writeln!(f, "\t{name} -> {val}")?;
                    }
                }
                Ok(())
            }

            ElrondResult::Result(r) => {
                for (spec, name) in r {
                    writeln!(f, "\t{name} : {spec}")?;
                }
                Ok(())
            }
        }
    }
}

/// See translate/translate.ml -> TenvEngine -> known_preds
#[allow(non_camel_case_types)]
#[derive(Serialize, Deserialize, PartialEq, Eq, Debug)]
pub enum ElrondKnownPredicate {
    length,
    sorted,
    member,
    head,
    order,
    once,
    left,
    right,
    para,
    ance,
    root,
}

impl Display for ElrondKnownPredicate {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            ElrondKnownPredicate::length => write!(f, "\"len\""),
            ElrondKnownPredicate::member => write!(f, "\"mem\""),
            ElrondKnownPredicate::head => write!(f, "\"hd\""),
            ElrondKnownPredicate::order => write!(f, "\"ord\""),
            ElrondKnownPredicate::left => write!(f, "\"left\""),
            ElrondKnownPredicate::right => write!(f, "\"right\""),
            ElrondKnownPredicate::sorted => write!(f, "\"sorted\""),
            ElrondKnownPredicate::once => write!(f, "\"once\""),
            ElrondKnownPredicate::para => write!(f, "\"para\""),
            ElrondKnownPredicate::ance => write!(f, "\"ance\""),
            ElrondKnownPredicate::root => write!(f, "\"root\""),
        }
    }
}

#[derive(Serialize, Deserialize, PartialEq, Debug)]
pub struct ElrondPredicates(pub Vec<ElrondKnownPredicate>);

impl Display for ElrondPredicates {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "let preds = [| {} |]",
            self.0.iter().map(ToString::to_string).join("; ")
        )
    }
}

#[derive(Debug)]
pub struct ElrondAssertion {
    pub preds: ElrondPredicates,
    pub pre_spec: Option<ElrondSpec>,
    pub post_spec: ElrondSpec,
}

impl Display for ElrondAssertion {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let ElrondAssertion {
            preds,
            pre_spec,
            post_spec,
        } = self;
        writeln!(f, "{preds}",)?;
        if let Some(pre_spec) = pre_spec {
            pre_spec.spec_writer_for_file(f, SpecKind::Pre)?;
        }
        writeln!(f)?;

        post_spec.spec_writer_for_file(f, SpecKind::Post)?;

        Ok(())
    }
}

enum SpecKind {
    Pre,
    Post,
}

impl Display for SpecKind {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            SpecKind::Pre => write!(f, "pre"),
            SpecKind::Post => write!(f, "post"),
        }
    }
}

impl ElrondSpec {
    fn spec_writer_for_file(
        &self,
        f: &mut std::fmt::Formatter<'_>,
        kind: SpecKind,
    ) -> std::fmt::Result {
        let ElrondSpec(vars, formula) = self;
        write!(
            f,
            "let {kind} {} =\n  {}\n",
            vars.iter()
                .map(ToString::to_string)
                .map(|s| "(".to_string() + &s + ")")
                .join(" "),
            formula
        )
    }
}
