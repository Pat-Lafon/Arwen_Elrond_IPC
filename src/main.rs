use arwen_elrond_ipc::{
    ipc::{ArwenMessage, ArwenSetup, ElrondKnownPredicate, ElrondPredicates},
    Elrond,
};

#[cfg(test)]
use arwen_elrond_ipc::ipc::{
    ElrondAssertion, ElrondForallFormula, ElrondPred, ElrondSimpleExpr, ElrondSpec, ElrondTpedvar,
    ElrondType,
};

fn main() {
    let mut e = Elrond::new();

    e.send_message(ArwenMessage::Start).unwrap();

    let dir = "ADT-Lemma-Discovery".to_string();
    // Source ocaml file
    let sourcefile = dir.clone() + "/data/customstk.ml";
    let assertionfile = dir + "/data/customstk_assertion1.ml";
    let outputdir = "customstk_out".to_string();

    // AKA the name of the function being verified
    let client_name = "concat".to_string();

    // fnames are the component names

    // This is a little tricky, you will get weird results if some of these are missing
    let predicates = ElrondPredicates(vec![
        ElrondKnownPredicate::member,
        ElrondKnownPredicate::head,
    ]);

    // todo build in a check such that only the list of provided predicates are used in any specifications

    e.send_message(ArwenMessage::Setup(ArwenSetup {
        sourcefile,
        assertionfile,
        outputdir,
        client_name,
        predicates,
    }))
    .unwrap();

    println!("{}", e.receive_message());

    e.kill().unwrap();
}

#[test]
fn test_customstk_assertion1() {
    // Read in the file called "ADT-Lemma-Discovery/data/customstk_assertion1.ml"
    let assertionfile = "ADT-Lemma-Discovery/data/customstk_assertion1.ml";
    let assertion = std::fs::read_to_string(assertionfile).unwrap();

    let generic_type = ElrondType::Generic("Customstk.t".to_string());

    let my_assertion = ElrondAssertion {
        preds: ElrondPredicates(vec![
            ElrondKnownPredicate::member,
            ElrondKnownPredicate::head,
        ]),
        pre_spec: None,
        post_spec: ElrondSpec(
            vec![
                ElrondTpedvar(generic_type.clone(), "l1".to_string()),
                ElrondTpedvar(generic_type.clone(), "l2".to_string()),
                ElrondTpedvar(generic_type.clone(), "l3".to_string()),
                ElrondTpedvar(ElrondType::Int, "u".to_string()),
            ],
            ElrondForallFormula(
                vec![],
                ElrondPred::And(vec![
                    ElrondPred::Iff(
                        Box::new(ElrondPred::Atom(ElrondSimpleExpr::Op(
                            ElrondType::Bool,
                            "mem".to_string(),
                            vec![
                                ElrondSimpleExpr::Var(generic_type.clone(), "l3".to_string()),
                                ElrondSimpleExpr::Var(ElrondType::Int, "u".to_string()),
                            ],
                        ))),
                        Box::new(ElrondPred::Or(vec![
                            ElrondPred::Atom(ElrondSimpleExpr::Op(
                                ElrondType::Bool,
                                "mem".to_string(),
                                vec![
                                    ElrondSimpleExpr::Var(generic_type.clone(), "l1".to_string()),
                                    ElrondSimpleExpr::Var(ElrondType::Int, "u".to_string()),
                                ],
                            )),
                            ElrondPred::Atom(ElrondSimpleExpr::Op(
                                ElrondType::Bool,
                                "mem".to_string(),
                                vec![
                                    ElrondSimpleExpr::Var(generic_type.clone(), "l2".to_string()),
                                    ElrondSimpleExpr::Var(ElrondType::Int, "u".to_string()),
                                ],
                            )),
                        ])),
                    ),
                    ElrondPred::Implies(
                        Box::new(ElrondPred::Atom(ElrondSimpleExpr::Op(
                            ElrondType::Bool,
                            "hd".to_string(),
                            vec![
                                ElrondSimpleExpr::Var(generic_type.clone(), "l3".to_string()),
                                ElrondSimpleExpr::Var(ElrondType::Int, "u".to_string()),
                            ],
                        ))),
                        Box::new(ElrondPred::Or(vec![
                            ElrondPred::Atom(ElrondSimpleExpr::Op(
                                ElrondType::Bool,
                                "hd".to_string(),
                                vec![
                                    ElrondSimpleExpr::Var(generic_type.clone(), "l1".to_string()),
                                    ElrondSimpleExpr::Var(ElrondType::Int, "u".to_string()),
                                ],
                            )),
                            ElrondPred::Atom(ElrondSimpleExpr::Op(
                                ElrondType::Bool,
                                "hd".to_string(),
                                vec![
                                    ElrondSimpleExpr::Var(generic_type, "l2".to_string()),
                                    ElrondSimpleExpr::Var(ElrondType::Int, "u".to_string()),
                                ],
                            )),
                        ])),
                    ),
                ]),
            ),
        ),
    };
    insta::assert_display_snapshot!(assertion, my_assertion);
}
