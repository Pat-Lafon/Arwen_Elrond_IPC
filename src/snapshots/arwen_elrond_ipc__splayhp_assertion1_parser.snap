---
source: src/main.rs
expression: parser.parse(&assertion).unwrap()
---
let preds = [| "mem" |]

let post (x : int) (tree1 : Splayhp.t) (tree2 : Splayhp.t) (tree3 : Splayhp.t) (u : int) =
  iff (mem tree1 u) ((mem tree2 u || mem tree3 u))

